pkg/products/threescale/reconciler.go

```diff
@@ -1533,26 +1535,62 @@ func (r *Reconciler) reconcileOpenshiftUsers(ctx context.Context, installation *
 		}
 	}
 
-	for _, kcUser := range added {
-		res, err := r.tsClient.AddUser(strings.ToLower(kcUser.UserName), strings.ToLower(kcUser.Email), "", *accessToken)
-
-		// when the failure of user happens we don't want to block the reconciler.
-		// failure to create a user can happen in the case of the username being too long
-		// the max allowed user length is 40 characters in 3scale.
-		// The reconciler will continue to allow the installation to happen and a metric
-		// will be exposed and alert fire to alert to the creation failure
-		metrics.SetThreeScaleUserAction(res.StatusCode, kcUser.UserName, http.MethodPost)
-		if err != nil {
-			r.log.Error(fmt.Sprintf("Failed to add keycloak user %s to 3scale", kcUser.UserName), err)
+	for _, tsUser := range updated {
+		if tsUser.UserDetails.Username != *systemAdminUsername {
+			genKcUser, err := getGeneratedKeycloakUser(ctx, serverClient, rhssoConfig.GetNamespace(), tsUser)
+
+			if err != nil {
+				r.log.Warning("Failed to getting generate keycloak user: " + err.Error())
+				continue
+			}
+
+			_, err = r.tsClient.UpdateUser(tsUser.UserDetails.Id, genKcUser.Spec.User.UserName, tsUser.UserDetails.Email, *accessToken)
+			if err != nil {
+				r.log.Warning("Failed to updating 3scale user details: " + err.Error())
+			}
 		}
-		if res.StatusCode != http.StatusCreated {
-			r.log.Error(fmt.Sprintf("Failed to add keycloak user %s to 3scale with status code %d", kcUser.UserName, res.StatusCode), errors.New("error on http request"))
+	}
+
+	for _, kcUser := range added {
+		user, _ := r.tsClient.GetUser(kcUser.UserName, *accessToken)
+		// recheck the user is new.
+		// 3scale user may being update during the update phase
+		if user == nil {
+			var statusCode int
+			res, err := r.tsClient.AddUser(strings.ToLower(kcUser.UserName), strings.ToLower(kcUser.Email), "", *accessToken)
+
+			if err != nil {
+				r.log.Error(fmt.Sprintf("Failed to add keycloak user %s to 3scale", kcUser.UserName), err)
+			}
+
+			if res != nil {
+				statusCode = res.StatusCode
+			} else {
+				statusCode = 503
+			}
```

Comment:
```suggestion
			var statusCode int
			res, err := r.tsClient.AddUser(strings.ToLower(kcUser.UserName), strings.ToLower(kcUser.Email), "", *accessToken)

			if err != nil {
				r.log.Error(fmt.Sprintf("Failed to add keycloak user %s to 3scale", kcUser.UserName), err)
				statusCode = http.StatusServiceUnavailable
			} else {
			       statusCode = res.StatusCode
			}
```
We could the statusCode stuff as part of the `err` check conditional block and use `http.StatusServiceUnavailable` constant instead of `503`. What do you think ? :thinking: